smolyay/smolyak.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Jul  5 08:00:15 2022
+
+@author: mzf0069
+"""
+
+import numpy
+import scipy.special
+
+
+class IndexGrid:
+    """Indexes of grid points for constructing a surrogate function.
+
+    Parameters
+    ----------
+    dimension : int
+        Number of independent variables.
+    exactness : int
+        Level of exactness of the approximation.
+    index_per_level : list
+        Number of indexes at each level.
+    """
+
+    def __init__(self, dimension, exactness, index_per_level):
+        self.dimension = dimension
+        self.exactness = exactness
+        self.index_per_level = index_per_level
+        self._needs_update = True
+
+    def _update(self):
+        """Update the indexes of grid points.
+
+        It generates the indexes for each level
+        depending on the index_per_level, and makes the
+        indexes of grid points.
+
+        """
+        # requirement for generating the grid points
+        if len(self.index_per_level) < self.exactness + 1:
+            raise IndexError(
+                "index_per_level must be an array with a"
+                " length of at least {}"
+                .format(self.exactness + 1))
+
+        # cumulative sum to get end index of each level, up to exactness+1
+        end_levels = numpy.cumsum(self.index_per_level[:self.exactness+1])
+        # create ranges of indexes at each level
+        level_indexes = [list(range(end-n, end))
+                         for end, n in zip(end_levels, self.index_per_level)]
+        self._level_indexes = level_indexes
+
+        points = None
+        # get all combinations of points at each level
+        for summ in range(self.dimension, self.exactness+self.dimension+1):
+            # use generate_compositions function which implements NEXCOM
+            compositions = generate_compositions(summ, self.dimension)
+            for composition in compositions:
```

Comment:
If you use the generator style I suggested, this will work:
```suggestion
            for composition in generate_compositions(summ, self.dimension, include_zeros=False):
```

Reply:
Done.